pkg/controller/job/job_controller.go

```diff
@@ -646,32 +724,240 @@ func (jm *Controller) syncJob(key string) (bool, error) {
+				errCh <- err
+				utilruntime.HandleError(err)
+			}
+		}(pods[i])
+	}
+	wg.Wait()
+	return successfulDeletes, errorFromChannel(errCh)
+}
+
 // deleteJobPods deletes the pods, returns the number of successful removals
 // and any error.
 func (jm *Controller) deleteJobPods(job *batch.Job, jobKey string, pods []*v1.Pod) (int32, error) {
 	errCh := make(chan error, len(pods))
 	successfulDeletes := int32(len(pods))
+
+	failDelete := func(pod *v1.Pod, err error) {
+		// Decrement the expected number of deletes because the informer won't observe this deletion
+		jm.expectations.DeletionObserved(jobKey)
+		if !apierrors.IsNotFound(err) {
+			klog.V(2).Infof("Failed to delete Pod", "job", klog.KObj(job), "pod", klog.KObj(pod), "err", err)
+			atomic.AddInt32(&successfulDeletes, -1)
+			errCh <- err
+			utilruntime.HandleError(err)
+		}
+	}
+
 	wg := sync.WaitGroup{}
 	wg.Add(len(pods))
 	for i := range pods {
 		go func(pod *v1.Pod) {
 			defer wg.Done()
+			if patch := removeTrackingFinalizerPatch(pod); patch != nil {
+				if err := jm.podControl.PatchPod(pod.Namespace, pod.Name, patch); err != nil {
+					failDelete(pod, fmt.Errorf("removing completion finalizer: %w", err))
+					return
+				}
+			}
 			if err := jm.podControl.DeletePod(job.Namespace, pod.Name, job); err != nil {
-				// Decrement the expected number of deletes because the informer won't observe this deletion
-				if jobKey != "" {
-					jm.expectations.DeletionObserved(jobKey)
+				failDelete(pod, err)
+			}
+		}(pods[i])
+	}
+	wg.Wait()
+	return successfulDeletes, errorFromChannel(errCh)
+}
+
+// removeTrackingFinalizersFromAllPods removes finalizers from any Job Pod. This is called
+// when Job tracking with finalizers is disabled.
+func (jm *Controller) removeTrackingFinalizersFromAllPods(pods []*v1.Pod) error {
+	var podsWithFinalizer []*v1.Pod
+	for _, pod := range pods {
+		if hasJobTrackingFinalizer(pod) {
+			podsWithFinalizer = append(podsWithFinalizer, pod)
+		}
+	}
+	if len(podsWithFinalizer) == 0 {
+		return nil
+	}
+	_, err := jm.removeTrackingFinalizerFromPods(podsWithFinalizer)
+	return err
+}
+
+// trackJobStatusAndRemoveFinalizers does:
+// 1. Add finished Pods to .status.uncountedTerminatedPods
+// 2. Remove the finalizers from the Pods if they completed or were removed
+//    or the job was removed.
+// 3. Increment job counters for pods that no longer have a finalizer.
+// 4. Add Complete condition if satisfied with current counters.
+// It does this in a controlled way such that the size of .status doesn't grow
+// too much.
+func (jm *Controller) trackJobStatusAndRemoveFinalizers(job *batch.Job, pods []*v1.Pod, succeededIndexes orderedIntervals, uncounted uncountedTerminatedPods, finishedCond *batch.JobCondition, needsFlush bool) error {
+	isIndexed := isIndexedJob(job)
+	var podsToRemoveFinalizer []*v1.Pod
+	uncountedStatus := job.Status.UncountedTerminatedPods
+	var newSucceededIndexes []int
+	flushUncountedAndRemoveFinalizers := func(last bool) error {
```

Comment:
Any particular reason for embedding this rather big function? I prefer to use function embedding only for small ones, it's easier to read w/o embedding, imo.

Reply:
Yeah, reading through the code it's unnecessarily complicated. I'd prefer this being separate method with clear inputs and outputs, the current approach makes it hard to follow the logic.